api/v1alpha1/playbookexecution_types.go

```diff
@@ -25,19 +25,30 @@ import (
 
 // PlaybookExecutionSpec defines the desired state of PlaybookExecution
 type PlaybookExecutionSpec struct {
-	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
-
-	// Foo is an example field of PlaybookExecution. Edit playbookexecution_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	ExecutionStrategy ExecutionStrategy `json:"executionStrategy"`
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Default=0
+	ExecutionAttempt uint8 `json:"executionAttempt,omitempty" description:"the number of times the playbook has been executed" default:"0"`
+	// RunnerEvents     *models.AnsibleRunnerJobEventYaml `json:"runnerEvents,omitempty"`
+	// Events results.AnsiblePlaybookJSONResults `json:"events,omitempty"`
```

Comment:
I think this is the right place to put the fields related to the execution results: am I right?

`RunnerEvent` is the output compliant with fleet management. `RunnerEvents` type is generated in Makefile `generate-messages` target (from `internal/ansible/schema/ansibleRunnerJobEvent.yaml`).

`Events` are the results produced by `ansible-playbook` command.
`Events` type is `results.AnsiblePlaybookJSONResults` and it should be imported from `github.com/apenella/go-ansible/pkg/stdoutcallback/results`.

The problem is that, when I add `Events results.AnsiblePlaybookJSONResults` then the `make manifests` target fails as follow:

```bash
> make manifests                                                                                                                                                                                                                                                                    
gojsonschema --yaml-extension yaml -p models internal/ansible/schema/ansibleRunnerJobEvent.yaml -o internal/ansible/models/runner-job-event-gen.go 
/home/gloria/repos/flotta-operator/vendor/github.com/apenella/go-ansible/pkg/stdoutcallback/results/JSONResults.go:16:20: unsupported AST kind *ast.InterfaceType
/home/gloria/repos/flotta-operator/vendor/github.com/apenella/go-ansible/pkg/stdoutcallback/results/JSONResults.go:17:20: unsupported AST kind *ast.InterfaceType
/home/gloria/repos/flotta-operator/vendor/github.com/apenella/go-ansible/pkg/stdoutcallback/results/JSONResults.go:93:19: unsupported AST kind *ast.InterfaceType
/home/gloria/repos/flotta-operator/vendor/github.com/apenella/go-ansible/pkg/stdoutcallback/results/JSONResults.go:94:30: map values must be a named type, not *ast.InterfaceType
/home/gloria/repos/flotta-operator/vendor/github.com/apenella/go-ansible/pkg/stdoutcallback/results/JSONResults.go:99:19: unsupported AST kind *ast.InterfaceType
Error: not all generators ran successfully
run `controller-gen crd:trivialVersions=true,preserveUnknownFields=false rbac:roleName=manager-role webhook paths=./... output:crd:artifacts:config=config/crd/bases -w` to see all available markers, or `controller-gen crd:trivialVersions=true,preserveUnknownFields=false rbac:roleName=manager-role webhook paths=./... output:crd:artifacts:config=config/crd/bases -h` for usage
make: *** [Makefile:70: manifests] Error 1
```

The `unsupported AST kind *ast.InterfaceType` are related to `interface{}` fields. AFAIU, `controller-gen` can't handle `interface{}`. 
Do you have any suggestions?

Reply:
Another question about `AnsibleRunnerJobEventYaml` - it has YAML in its name, but it is generated with json tags. Also, adding linter to verify the yaml can be useful.

Regarding the error with controller-gen, there is a discussion (linked inside) and solution for it here: https://github.com/kubernetes-sigs/controller-tools/issues/519#issuecomment-726489250
This will require to do few changes on go-ansible side.

Reply:
> Execution results can go into status or into events.

AFAIK, the status should not "store" data and could be generated on the fly at any moment. 
"The status summarizes the current state of the object in the system, and is usually persisted with the object by automated processes but may be generated on the fly. As a general guideline, fields in status should be the most recent observations of actual state, but they may contain information such as the results of allocations or similar operations which are executed in response to the object's spec."  ([API Conventions - Spec and Status](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status) )

So, if it's ok for you, I'll use Events.

> AnsibleRunnerJobEventYaml - it has YAML in its name

The name is auto generated and at the moment `go-jsonschema` doesn't allow customization. It basically transforms filename+extension  in its CamelCase (`ansibleRunnerJobEvent.yaml`  ->  `AnsibleRunnerJobEventYaml` )

> Also, adding linter to verify the yaml can be useful

Good idea! I'll add it.

> Regarding the error with controller-gen....

I think that using Events solves the problem since they don't have to be specified in CRD but in the `swagger.yaml` (correct?)

> This will require to do few changes on go-ansible side

Question to know your point of view: I saw that post but do you think an external library that legitimately uses the `interface{}` type can be changed to be `controller-gen` compliant? Is that a good enough motivation?

Reply:
A library that 

> > Execution results can go into status or into events.
> 
> AFAIK, the status should not "store" data and could be generated on the fly at any moment. "The status summarizes the current state of the object in the system, and is usually persisted with the object by automated processes but may be generated on the fly. As a general guideline, fields in status should be the most recent observations of actual state, but they may contain information such as the results of allocations or similar operations which are executed in response to the object's spec." ([API Conventions - Spec and Status](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status) )

I was looking for this exact link :-) 

> 
> So, if it's ok for you, I'll use Events.
> 

+1

> > AnsibleRunnerJobEventYaml - it has YAML in its name
> 
> The name is auto generated and at the moment `go-jsonschema` doesn't allow customization. It basically transforms filename+extension in its CamelCase (`ansibleRunnerJobEvent.yaml` -> `AnsibleRunnerJobEventYaml` )
> 

thanks for clarifying 

> > Also, adding linter to verify the yaml can be useful
> 
> Good idea! I'll add it.
> 
> > Regarding the error with controller-gen....
> 
> I think that using Events solves the problem since they don't have to be specified in CRD but in the `swagger.yaml` (correct?)
> 

That should be correct. Events are managed by their own schema that we'll use.

> > This will require to do few changes on go-ansible side
> 
> Question to know your point of view: I saw that post but do you think an external library that legitimately uses the `interface{}` type can be changed to be `controller-gen` compliant? Is that a good enough motivation?

When we consume  library that doesn't satisfy all of our needs the options are as follow:
1. Open an issue that describe the problem and offer a PR to address it if agreed by the maintainers. 
2. Adapt our code.
3. Don't use it (seek for an alternative or develop own).
4. Fork it and adapt it.

The first option might lead to breaking the API of the library. If there is a good reasoning for doing it, then only in the next major version of that library the changes will be released.
But prior to it, reporting an issue with description of the problem can lead to the maintainer/member of the project to offer an alternative way or a remedy to the situation.

There need to be good reasoning for changing an API, but as I see it, the value of a library is to satisfy its consumers needs while making sense(this is too wide definition, but basically, not to damage dev principles :-)). A library without consumers is just a tool that will not be used.
If there is a majority of consumers for a library, than it will be hard to push a change that can affect multiple users.

The legitimacy will be derived from the consensus around that change.
We used to maintain libraries with only known 2 consumers other than us and it was easy to push changes and release frequently as requirements arrived.

And sometimes we'll find our self seeking for a very needed change requested by multiple users and the maintainers will not agree to. So the options are 2-4 which need to be well evaluated.

Reply:
Thanks @masayag!